tracer/src/Datadog.Trace/Debugger/Symbols/SymbolSink.cs

```diff
@@ -0,0 +1,132 @@
+        {
+            _alreadyProcessed = new HashSet<string>();
+            _environment = environment;
+            _serviceVersion = serviceVersion;
+            _serviceName = serviceName;
+            _symbolUploader = uploader;
+            _symbolExtractor = symbolExtractor;
+            _assemblySemaphore = new SemaphoreSlim(1);
+            _cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public static SymbolSink Create(string? environment, string? serviceVersion, string serviceName, SymbolUploader uploader, SymbolExtractor symbolExtractor)
+        {
+            return new SymbolSink(environment, serviceVersion, serviceName, uploader, symbolExtractor);
+        }
+
+        private void RegisterToAssemblyLoadEvent()
+        {
+            AppDomain.CurrentDomain.AssemblyLoad += async (_, args) =>
+            {
+                await ProcessItemAsync(args.LoadedAssembly).ConfigureAwait(false);
+            };
+        }
+
+        private async Task ProcessItemAsync(Assembly assembly)
+        {
+            await _assemblySemaphore.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                await ProcessItem(assembly).ConfigureAwait(false);
+            }
+            finally
+            {
+                _assemblySemaphore.Release();
+            }
+        }
+
+        private async Task ProcessItem(Assembly assembly)
+        {
+            try
+            {
+                if (!_alreadyProcessed.Add(assembly.Location))
+                {
+                    return;
+                }
+
+                await GetAssemblySymbols(assembly).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Error while trying to extract assembly symbol {Assembly}", assembly);
+            }
+        }
+
+        private async Task GetAssemblySymbols(Assembly assembly)
+        {
+            if (AssemblyFilter.ShouldSkipAssembly(assembly))
+            {
+                return;
+            }
+
+            var assemblyScope = _symbolExtractor.GetAssemblySymbol(assembly);
+
+            var root = new Model.Root
+            {
+                Service = _serviceName,
+                Env = _environment,
+                Language = "dotnet",
+                Version = _serviceVersion,
+                Scopes = new[] { assemblyScope }
+            };
+
+            foreach (var classSymbol in _symbolExtractor.GetClassSymbols(assembly))
+            {
+                if (classSymbol == default)
+                {
+                    continue;
+                }
+
+                root.Scopes[0].Scopes.Add(classSymbol);
+                await _symbolUploader.SendSymbol(root).ConfigureAwait(false);
+                root.Scopes[0].Scopes.Clear();
+            }
+        }
+
+        public async Task StartExtractingAssemblySymbolsAsync()
+        {
+            RegisterToAssemblyLoadEvent();
```

Comment:
1. We should be careful not to do _any_ work at all or even register to assembly load event if the aforementioned environment variable is turned off, or if `DiscoveryService` tells us the SymDB endpoint is not available. 
2. As far as I can tell this PR doesn't address the requirement that Symbol Database upload should only occur if we receive an instruction to perform it via Remote Configuration. If your plan it add that in a subsequent PR, since we don't want it on by default, I propose it should be gated by an environment variable that should be `false` by default for the time being, i.e. something like `DD_INTERNAM_FORCE_SYMBOL_DATABASED_UPLOAD`?